readme.md
<!--
  Created at: 07/08/2021 15:04:13 Thursday
  Modified at: 08/05/2021 01:39:37 PM Thursday
-->

# safeWriting

Write in websites without being tracked every keystroke

## Usage

Just click on a typing area.

## Installation

1. Install a userscript manager like Violentmonkey
2. Open [build](https://gitea.com/thisago/safeWriting/raw/branch/master/build/safeWriting.user.js) on your browser
3. Confirm installation

## License

MIT


changelog.md
<!--
  Created at: 07/08/2021 15:47:38 Thursday
  Modified at: 09/17/2021 10:48:49 PM Friday
-->

# Changelog

## Version 0.3.0 (09/17/2021 22:48:26 Friday)

- changed event from `click` to `dblclick`
- Added description
- Changed comment header of all files

### Allowed

- textarea
- input
  - type=time
  - type=week
  - type=range
  - type=month
  - type=color
  - type=date
  - type=datetime
  - disabled
  - readonly

---

## Version 0.2.0 (07/08/2021 17:21:25 Thursday)

- Removed blur element before entering the text
- Temporary removed textarea trigger (issue [#1](https://gitea.com/thisago/safeWriting/issues/1))
- Simplified the description to not need to translate (in another languages like the name)
- Added @downloadURL metadata
- Removed the trigger on focus because if element has autofocus it will be
  triggered automatically (just clicking)
- Removed the trigger for disabled and readonly inputs
- Added more inputs into test
- Removed useless variables in javascript of test page
- Added colors to the keys in test page
- Added history with
  - Configurable command
  - Enabling/disabling
  - Clean history


test/script.js
/*
  Created at: 07/05/2021 19:28:00 Monday
  Modified at: 07/08/2021 04:14:15 PM Thursday
*/

const maxLogSize = 20

const log = document.getElementById("log")

/**
 * Binds the key press listener on element
 *
 * @param {HTMLElement} el
 */
function bindListener(el) {
  let pressedKeys = {}
  el.addEventListener("keydown", ev => {
    const key = ev.key
    if (!pressedKeys[key]) // if start time not setted
      pressedKeys[key] = Date.now() // set the pressed time
  })
  el.addEventListener("keyup", ev => {
    const
      key = ev.key,
      lastTime = pressedKeys[key] // get the time of keydown
    if (!lastTime) return
    pressedKeys[key] = undefined // delete the time of keydown

    const
      pressedTime = Date.now() - lastTime,
      text = document.createElement("div")

    text.innerHTML = `Key '<span style="--i:${key.charCodeAt(0) * 30}">${key}</span>' was pressed for <span style="--i:${pressedTime}">${pressedTime}ms</span>`

    log.append(text)

    while (log.childNodes.length > maxLogSize)
      log.childNodes[0].remove()
  })
}

bindListener(document.body)


src/safeWriting.nim
#[
  Created at: 07/05/2021 18:45:00 Monday
  Modified at: 09/17/2021 09:32:51 PM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

{.experimental: "codeReordering".}

import std/[jsffi, dom]
from std/strutils import join
from std/strformat import fmt
import pkg/gm_api

import ./safeWriting/defs
import ./safeWriting/header
import std/asyncjs

{.emit: userscriptHeader.}

when isMainModule:
  discard main()

proc blur*(el: Node) {.importcpp.}
proc addEventListener[T](et: EventTarget; ev: cstring; cb: proc (ev: Event): T) {.importcpp.}

proc main {.async.} =
  let
    historyEnabled = (await GM.getValue("history_enabled", "1")) == "1"
    historyCommand = await GM.getValue("history_command", "@")

  proc handleInput(ev: Event) {.async.} =
    if ev.isTrusted:
      let
        el = ev.target
        text = window.prompt(el.getAttribute "placeholder", el.value)
      if not text.isNull:
        block:
          if historyEnabled:
            if text == historyCommand:
              el.value = await GM.getValue("history_value", "")
              break
            await GM.setValue("history_value", text)
          el.value = text

  for el in document.querySelectorAll inputElements.join ",":
    for event in inputTriggerEvents:
      el.addEventListener(event, handleInput)

  GM.registerMenuCommand(
    (if historyEnabled: "Disable" else: "Enable") & " History",
    proc () =
      discard GM.setValue("history_enabled", if historyEnabled: "0" else: "1")
      window.alert "Success! Please reload the page"
    ,
    "h"
  )
  GM.registerMenuCommand(
    cstring fmt"Change history command ({historyCommand})",
    proc () =
      let text = window.prompt("History command", "")
      if not text.isNull and text != "":
        discard GM.setValue("history_command", text)
        window.alert "Command successfully changed. Reload the page to use"
    ,
    "c"
  )
  GM.registerMenuCommand(
    cstring "Clean saved history",
    proc () =
      discard GM.setValue("history_value", "")
      window.alert "Successfully cleaned!"
    ,
    "c"
  )


src/safeWriting/header.nim
#[
  Created at: 07/08/2021 14:01:57 Thursday
  Modified at: 09/17/2021 09:37:25 PM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

import pkg/gm_api/metadata

const userscriptHeader* = genMetadataBlock(
  name = "Safe Writing",
  nameTranslations = {
    "pt": "Escrita segura",
    "es": "Escritura segura",
    "it": "Scrittura sicura",
  },
  grant = [
    GmPermitions.getValue,
    GmPermitions.setValue,
    GmPermitions.registerMenuCommand,
  ],
  author = "Thiago Navarro",
  version = "0.3.0",
  runAt = GmRunAt.docIdle,
  downloadUrl = "https://raw.githubusercontent.com/thisago/safeWriting/master/build/safeWriting.user.js",
  description = "Write in websites without being tracked every keystroke",
  homepageUrl = "https://github.com/thisago/safeWriting",
)


src/safeWriting/defs.nim
#[
  Created at: 07/08/2021 13:03:35 Thursday
  Modified at: 09/17/2021 09:44:45 PM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

from std/strutils import join

const inputElements* = [
  "input" & [
    ":not([type=button])",
    ":not([type=checkbox])",
    # ":not([type=color])",
    # ":not([type=date])",
    # ":not([type=datetime])",
    ":not([type=file])",
    ":not([type=hidden])",
    ":not([type=image])",
    # ":not([type=month])",
    ":not([type=radio])",
    # ":not([type=range])",
    ":not([type=reset])",
    ":not([type=submit])",
    # ":not([type=time])",
    # ":not([type=week])",
    # ":not([disabled])",
    # ":not([readonly])",
  ].join(""),
  "textarea", # Issue #1
  "[contenteditable=true]"
]
const inputTriggerEvents* = [
  "dblclick",
]
